roslibrust/src/ros1/subscriber.rs

```diff
@@ -0,0 +1,219 @@
+            phantom: PhantomData,
+        }
+    }
+
+    pub async fn next(&mut self) -> Result<T, Box<dyn std::error::Error>> {
+        let data = self.receiver.recv().await.map_err(|err| Box::new(err))?;
+        Ok(serde_rosmsg::from_slice(&data[..]).map_err(|err| Box::new(err))?)
+    }
+}
+
+pub struct Subscription {
+    subscription_tasks: Vec<ChildTask<()>>,
+    _msg_receiver: broadcast::Receiver<Vec<u8>>,
+    msg_sender: broadcast::Sender<Vec<u8>>,
+    connection_header: ConnectionHeader,
+    known_publishers: Arc<RwLock<Vec<String>>>,
+}
+
+impl Subscription {
+    pub fn new(
+        node_name: &str,
+        topic_name: &str,
+        topic_type: &str,
+        queue_size: usize,
+        msg_definition: String,
+        md5sum: String,
+    ) -> Self {
+        let (sender, receiver) = broadcast::channel(queue_size);
+        let connection_header = ConnectionHeader {
+            caller_id: node_name.to_owned(),
+            latching: false,
+            msg_definition,
+            md5sum,
+            topic: topic_name.to_owned(),
+            topic_type: topic_type.to_owned(),
+            tcp_nodelay: false,
+        };
+
+        Self {
+            subscription_tasks: vec![],
+            _msg_receiver: receiver,
+            msg_sender: sender,
+            connection_header,
+            known_publishers: Arc::new(RwLock::new(vec![])),
+        }
+    }
+
+    pub fn topic_type(&self) -> &str {
+        self.connection_header.topic_type.as_str()
+    }
+
+    pub fn get_receiver(&self) -> broadcast::Receiver<Vec<u8>> {
+        self.msg_sender.subscribe()
+    }
+
+    pub async fn add_publisher_source(
+        &mut self,
+        publisher_uri: &str,
+    ) -> Result<(), std::io::Error> {
+        let is_new_connection = {
+            self.known_publishers
+                .read()
+                .await
+                .iter()
+                .find(|publisher| publisher.as_str() == publisher_uri)
+                .is_none()
+        };
+
+        if is_new_connection {
+            let node_name = self.connection_header.caller_id.clone();
+            let topic_name = self.connection_header.topic.clone();
+            let connection_header = self.connection_header.clone();
+            let sender = self.msg_sender.clone();
+            let publisher_list = self.known_publishers.clone();
+            let publisher_uri = publisher_uri.to_owned();
+
+            let handle = tokio::spawn(async move {
+                if let Ok(mut stream) = establish_publisher_connection(
+                    &node_name,
+                    &topic_name,
+                    &publisher_uri,
+                    connection_header,
+                )
+                .await
+                {
+                    publisher_list.write().await.push(publisher_uri.to_owned());
+                    // Repeatedly read from the stream until its dry
+                    let mut read_buffer = Vec::with_capacity(4 * 1024);
```

Comment:
Something feels fishy here, and maybe I'm just not clear on the big picture.

It seems like a single call to read_buf could give a partial message, or multiple messages. From what I can tell whatever we shove down read_buf will attempted to be parsed as a single message, so i think there is a bug there?

I think we need to parse the ros header here, and packetize off of message length? Am I crazzy?

Reply:
No, not crazy, that's totally possible. We're a little bit reliant on the other side of the stream sending valid data, but I do think we have to assume that and attempt to read the length to intelligently consume from the buffer. I'm not certain about the type and size of the buffer I'm using here. It's way overkill in a lot of cases, but it might add latency if a large single message got sent down the stream. On the other hand, I do think we want some kind of maximum in case a misbehaving node sends invalid data and we get tricked into trying to consume u32::MAX bytes or something.

Reply:
I'm inclined to say this works for small messages and we should fix this in a follow up MR.